src/main/java/com/google/api/gax/grpc/PageStreamingCallable.java

```diff
@@ -60,62 +66,115 @@ public String toString() {
+    @Override
+    public PageAccessor<ResourceT> getNextPage() {
+      Object nextToken = getNextPageToken();
+      if (nextToken == null) {
+        return null;
+      } else {
+        RequestT nextRequest =
+            pageDescriptor.injectToken(context.getRequest(), getNextPageToken());
+        return new PageAccessorImpl<>(callable,
+                                      pageDescriptor,
+                                      context.withRequest(nextRequest));
       }
+    }
 
-      @Override
-      public boolean hasNext() {
-        if (currentIter.hasNext()) {
-          return true;
-        }
-        if (nextRequest == null) {
-          return false;
-        }
-        ResponseT newPage =
-            Futures.getUnchecked(callable.futureCall(context.withRequest(nextRequest)));
-        currentIter = pageDescriptor.extractResources(newPage).iterator();
+    @Override
+    public String getNextPageToken() {
+      Object nextToken = pageDescriptor.extractNextToken(getPage());
+      if (nextToken == null || nextToken.equals(pageDescriptor.emptyToken())) {
+        return null;
+      } else {
+        return nextToken.toString();
+      }
+    }
 
-        Object nextToken = pageDescriptor.extractNextToken(newPage);
-        if (nextToken.equals(pageDescriptor.emptyToken())) {
-          nextRequest = null;
-        } else {
-          nextRequest = pageDescriptor.injectToken(nextRequest, nextToken);
-        }
-        return currentIter.hasNext();
+    private ResponseT getPage() {
+      if (currentPage == null) {
+        currentPage = getUnchecked(callable.futureCall(context));
       }
+      return currentPage;
+    }
 
-      @Override
-      public ResourceT next() {
-        if (!hasNext()) {
-          throw new NoSuchElementException();
-        }
-        return currentIter.next();
+    private class PageIterator extends AbstractIterator<ResourceT> {
+      private RequestT nextRequest;
+      private Iterator<ResourceT> currentIterator;
+
+      private PageIterator(RequestT request) {
+        nextRequest = request;
+        currentIterator = Collections.emptyIterator();
       }
 
       @Override
-      public void remove() {
-        throw new UnsupportedOperationException();
+      protected ResourceT computeNext() {
+        if (currentIterator.hasNext()) {
+          return currentIterator.next();
+        } else if (nextRequest == null) {
+          return endOfData();
+        } else {
+          ResponseT newPage =
+              getUnchecked(callable.futureCall(context.withRequest(nextRequest)));
+          Object nextToken = pageDescriptor.extractNextToken(newPage);
+          if (nextToken.equals(pageDescriptor.emptyToken())) {
+            nextRequest = null;
+          } else {
+            nextRequest = pageDescriptor.injectToken(nextRequest, nextToken);
+          }
+          currentIterator = pageDescriptor.extractResources(newPage).iterator();
+          return computeNext();
```

Comment:
Empty result will cause nextRequest to be null, which will call endOfData() in the next recursive call.